sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,808 @@
+            X = check_array(X, dtype=np.object)
+        else:
+            X = X_temp
+
+        _, n_features = X.shape
+        X_int = np.zeros_like(X, dtype=np.int)
+        X_mask = np.ones_like(X, dtype=np.bool)
+
+        for i in range(n_features):
+            Xi = X[:, i]
+            valid_mask = np.in1d(Xi, self.categories_[i])
+
+            if not np.all(valid_mask):
+                if handle_unknown == 'error':
+                    diff = np.unique(X[~valid_mask, i])
+                    msg = ("Found unknown categories {0} in column {1}"
+                           " during transform".format(diff, i))
+                    raise ValueError(msg)
+                else:
+                    # Set the problematic rows to an acceptable value and
+                    # continue `The rows are marked `X_mask` and will be
+                    # removed later.
+                    X_mask[:, i] = valid_mask
+                    Xi = Xi.copy()
+                    Xi[~valid_mask] = self.categories_[i][0]
+            X_int[:, i] = self._label_encoders_[i].transform(Xi)
+
+        return X_int, X_mask
+
+
+class OneHotEncoder(_BaseEncoder):
+    """Encode categorical integer features as a one-hot numeric array.
+
+    The input to this transformer should be an array-like of integers or
+    strings, denoting the values taken on by categorical (discrete) features.
+    The features are encoded using a one-hot (aka 'one-of-K' or 'dummy')
+    encoding scheme. This creates a binary column for each category and
+    returns a sparse matrix or dense array.
+    By default, encoder derives the categories based on the unique values
+    in each feature.  Alternatively, the input to this transformer can be a
+    matrix of already encoded integers denoting the values taken on by
+    categorical (discrete) features. In that case you have to specify
+    ``encoded_input=True`` and it is assumed that input features take on
+    values in the range [0, max(values)).
+
+    This encoding is needed for feeding categorical data to many scikit-learn
+    estimators, notably linear models and SVMs with the standard kernels.
+
+    Note: a one-hot encoding of y labels should use a LabelBinarizer
+    instead.
+
+    Read more in the :ref:`User Guide <preprocessing_categorical_features>`.
+
+    Parameters
+    ----------
+    categories : 'auto' or a list of lists/arrays of values.
+        Categories (unique values) per feature:
+
+        - 'auto' : Determine categories automatically from the training data.
+        - list : ``categories[i]`` holds the categories expected in the ith
+          column. The passed categories must be sorted and should not mix
+          strings and numeric values.
+
+        The used categories can be found in the ``categories_`` attribute.
+
+    encoded_input : boolean
+        How to interpret the input data:
+
+        encoded_input=False : categorical features that still need to be
```

Comment:
Don't use ` : descr`. The correct syntax for definition lists is:

```
encoded_input=False
    categorical features ...
```

I think the default should be 'auto', not None, with the default changing to False in two versions.

Do you think this option will be useful (for efficiency, for error handling, or for quality assurance) after the deprecation is finished? If not, should it just be called `legacy_mode`?

Does this setting affect the handling of `n_values`? Can  one use `encoded_input=True` and `categories` together?

This seems to mostly affect the handling of errors, in that even if a column isn't active in input, that value is not considered unknown....?